code/Control/UserDefinedFormController.php

```diff
@@ -620,87 +615,84 @@ public function confirmfolderform()
-            "new" => _t(__CLASS__.'.FOLDER_OPTIONS_NEW', 'Create a new folder (recommended)'),
-            "existing" => _t(__CLASS__.'.FOLDER_OPTIONS_EXISTING', 'Use an existing folder')
-        ], "new");
-        $fields->push($options);
-
-
-        $treeView = TreeDropdownField::create(
-            'FolderID',
-            '',
-            Folder::class
-        )->setValue($folder->ID);
-        $treeView->addExtraClass('pt-1');
-        $treeView->setDescription(EditableFileField::getFolderPermissionString($folder));
-        $fields->push($treeView);
-
-
-        $fields->push(HiddenField::create('ID', 'ID', $editableFormField->ID));
-
-        $submitAction = FormAction::create('confirmfolder', _t(__CLASS__.'.FORM_ACTION_CONFIRM', 'Save and continue'));
-        $submitAction->setUseButtonTag(false);
-        $submitAction->addExtraClass('btn');
-        $submitAction->addExtraClass('btn-primary');
-
-        $cancelAction = FormAction::create("cancel", _t('SilverStripe\\CMS\\Controllers\\CMSMain.Cancel', "Cancel"))
-            ->addExtraClass('btn-secondary')
-            ->setUseButtonTag(true);
-
-        $form = Form::create($this, 'ConfirmFolderForm', $fields, FieldList::create($submitAction, $cancelAction));
-        $form->setFormAction("UserDefinedFormController/confirmfolder");
-        $form->addExtraClass('form--no-dividers');
-
+        $form = $this->buildConfirmFolderForm(
+            $userForm->Title ?: '',
+            EditableFileField::getFolderPermissionString($folder)
+        );
+        $form->loadDataFrom(['FolderID' => $folderId, 'ID' => $id]);
 
+        // Convert the EditableFormField to an EditableFileField if it's not already one.
         if (!$editableFormField instanceof EditableFileField) {
             $editableFormField = $editableFormField->newClassInstance(EditableFileField::class);
             $editableFormField->write();
         }
-        $treeView->setSchemaData([
-            'data' => [
-                'urlTree' => "admin/pages/edit/EditForm/$userFormID/field/Fields/item/$id/ItemEditForm/field/FolderID/tree"
-            ]
-        ]);
-
 
         // create the schema response
         $parts = $this->getRequest()->getHeader(LeftAndMain::SCHEMA_HEADER);
         $schemaID = $this->getRequest()->getURL();
-        $data = FormSchema::singleton()
-            ->getMultipartSchema($parts, $schemaID, $form);
+        $data = FormSchema::singleton()->getMultipartSchema($parts, $schemaID, $form);
 
         // return the schema response
         $response = HTTPResponse::create(json_encode($data));
         $response->addHeader('Content-Type', 'application/json');
         return $response;
     }
 
+    public function ConfirmFolderForm(): Form
+    {
+        return $this->buildConfirmFolderForm();
+    }
+
+    private function buildConfirmFolderForm(string $suggestedFolderName = '', string $permissionFolderString = ''): Form
```

Comment:
I've moved all the Form Field logic to a dedicated method because I need a way to call it in `confirmfolderformschema` and `ConfirmFolderForm`. I would have called `ConfirmFolderForm` directly from `confirmfolderformschema`, but there's some form fields that get customised based on the field ID.